### PR TITLE
Gracefully handle HTTP build cache services rejecting too-large entries

### PR DESCRIPTION
**Pull request by ldaley**
Previously, servers rejecting the request with a 413 response and early termination caused an error to be displayed and no further load/store to the remote cache. Now, it is an accepted outcome and the the cache can continue to be used.

### CODE REVIEW

**Review by lptr — COMMENTED** [review]
I think it would be good to split out moving to the async HTTP implementation from this PR first to reduce the size of this PR.

Then we could introduce the new public API in a separate one. I'm leaning towards a new type, perhaps called `RemoteBuildCacheService`, see comments inline.

I see this implementation still wants to replace the stream-based protocol with a file-based one. As we discussed before, we should not lock us into having one file per cache artifact, as there are significant gains in implementing a more compact storage mechanism for cache artifacts.

File: subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheService.java [hunk 1]
```diff
@@ -56,7 +58,53 @@
...
+    /**
+     * Potentially stores an entry in the cache.
+     *
+     * @param key the cache key.
+     * @param writer the writer to write the data corresponding to the cache key.
+     * @throws BuildCacheException if the cache fails to store a cache entry for the given key
+     * @since 7.2
+     */
+    @Incubating
+    default StoreOutcome maybeStore(BuildCacheKey key, BuildCacheEntryWriter writer) throws BuildCacheException {
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/core-api/src/main/java/org/gradle/caching/BuildCacheService.java:104]: Yeah, I think that a new `RemoteBuildCacheService` is a much better direction. Will go that way.

File: subprojects/build-cache-http/build.gradle.kts [hunk 2]
```diff
@@ -14,6 +14,7 @@ dependencies {
     implementation(libs.slf4jApi)
     implementation(libs.guava)
     implementation(libs.commonsHttpclient)
+    implementation(libs.commonsHttpclientAsync)
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/build-cache-http/build.gradle.kts:17]: Will check.

**Comment by ldaley** [comment]
> I think it would be good to split out moving to the async HTTP implementation from this PR first to reduce the size of this PR.
> 
> Then we could introduce the new public API in a separate one. I'm leaning towards a new type, perhaps called `RemoteBuildCacheService`, see comments inline.
> 
> I see this implementation still wants to replace the stream-based protocol with a file-based one. As we discussed before, we should keep not lock us into having one file per cache artifact, as there are significant gains in implementing a more compact storage mechanism for cache artifacts.

We first need to decide if we want to retain the ability to zero-copy (i.e. transfer from socket to file and vice versa at the kernel instead of going through user space). It's going to be more “efficient”. By how much and whether it's noticeable would need dedicated analysis to determine. WRT usage of Apache async HTTP client, it's also the path of least resistance.

If we do want to do this, then the new service API has to work [Readable/Writable]ByteChannel. This would be feasible to do without locking in to file per artifact, as channels can be used to read/write to portions of a file. Implementations wanting to zero copy would have to type test the channel to see if it's a `FileChannel` and then use the zero-copy fast path, while retaining a non FileChannel slow path. Alternatively, we do predicate the API on FileChannel, but not on exclusive use of files. In practical terms, this would mean passing around a FileChannel and an offset/length to be used for reading and writing.

In inconclusive on the way to go forward. It seems a shame to not utilize zero copy when it's so close, but it does make things more complicated if we want to avoid baking FileChannel into the API.

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/ExceptionChainNoTraceRenderer.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+package org.gradle.caching.internal.controller.service;
+
+import java.io.PrintWriter;
+
+/**
+ * Mimics Throwable.printStackTrace(), but omits frames.
+ */
+// TODO - where does this live?
```
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/ExceptionChainNoTraceRenderer.java:24]: To where though? Any advice?

[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/ExceptionChainNoTraceRenderer.java:24]: How about `org.gradle.internal.exceptions` in `:base-services`?

**Comment by lptr** [comment]
> > I think it would be good to split out moving to the async HTTP implementation from this PR first to reduce the size of this PR.
> > Then we could introduce the new public API in a separate one. I'm leaning towards a new type, perhaps called `RemoteBuildCacheService`, see comments inline.
> > I see this implementation still wants to replace the stream-based protocol with a file-based one. As we discussed before, we should keep not lock us into having one file per cache artifact, as there are significant gains in implementing a more compact storage mechanism for cache artifacts.
> 
> We first need to decide if we want to retain the ability to zero-copy (i.e. transfer from socket to file and vice versa at the kernel instead of going through user space). It's going to be more “efficient”. By how much and whether it's noticeable would need dedicated analysis to determine. WRT usage of Apache async HTTP client, it's also the path of least resistance.

I don't have a deep enough understanding of the intricacies of the various HTTP client implementations, but it seems reasonable that we switch if the benefit is to have zero-copy. I would just do the switch in a separate PR first, and then worry about the change in API. Especially if we keep around the old `BuildCacheService`, we'll need to make it work with async HTTP as well.

> If we do want to do this, then the new service API has to work [Readable/Writable]ByteChannel. This would be feasible to do without locking in to file per artifact, as channels can be used to read/write to portions of a file. Implementations wanting to zero copy would have to type test the channel to see if it's a `FileChannel` and then use the zero-copy fast path, while retaining a non FileChannel slow path. Alternatively, we do predicate the API on FileChannel, but not on exclusive use of files. In practical terms, this would mean passing around a FileChannel and an offset/length to be used for reading and writing.
> 
> In inconclusive on the way to go forward. It seems a shame to not utilize zero copy when it's so close, but it does make things more complicated if we want to avoid baking FileChannel into the API.

I don't see much of a contradiction here. The internal build cache machinery requires restartable/random access to the bytes that make up the cached artifact. We can capture this in a new custom API that could replace `BuildCacheEntry[Reader/Writer]`. And then we can have different implementations backing that API using a `File` or a `FileChannel` or whatever else we might have, depending on how we implement our local cache.

**Comment by lptr** [comment]
Doing this is quite significant work, though, as it involves the redefinition of the whole public build cache implementation API. While I would very much like to spend time improving that API, we haven't planned to do that in the foreseeable future. I would also not like to make large-scale changes right now without thinking them through in the context of future plans for the local cache.

I wonder how far we can get right now by exposing a more flexible internal API to `HttpBuildCacheService` to make it possible to address your use case, and leave the larger scale restructuring to a later time.

**Comment by ldaley** [comment]
@lptr based on this, here is my plan:

- Add a new internal/unstable API that is based on ByteChannel and allows cache connectors to “not store”. 
- Don't touch any existing HttpClient stuff, just use async http client in the build-cache-http project
- Retain gracefully handling a 413-then-close response from the server as a “no store”

The prize we are after here is that trying to submit something too-large does not disable the cache for the rest of the build, and that's the most minimal way to do it.

File: subprojects/build-cache-http/build.gradle.kts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ldaley) [reply subprojects/build-cache-http/build.gradle.kts:16]: there can only be one

File: subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/DefaultBuildCacheEntryInternal.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        this.file = file;
+    }
+
+    @Override
+    public File getFile() {
+        return file;
+    }
+
+    @Override
+    public void markDownloading() {
```
Comment (ldaley) [inline subprojects/build-cache/src/main/java/org/gradle/caching/internal/controller/service/DefaultBuildCacheEntryInternal.java:37]: Having this here is quite weird. I'm trying to come up with something better, but haven't had any success.

**Comment by ldaley** [comment]
@lptr this is ready for review. No public API has changed.

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@bot-gradle test rfm

**Comment by ldaley** [comment]
@bot-gradle rfm

**Comment by lptr** [comment]
What is the next step here?

**Comment by etiennestuder** [comment]
@pioterj This should not get closed. What needs to happen to get this merged? Or is there no more intention to get this merged?

**Comment by pioterj** [comment]
@etiennestuder When we discussed this two months ago, @ldaley said he won't be able to revisit until this year. He wasn't sure what was left and needed some time to load up the context.

**Comment by ldaley** [comment]
ACK. Back on my radar.

**Comment by pioterj** [comment]
We still intent to look into it

**Comment by etiennestuder** [comment]
Please keep this PR open.

**Comment by etiennestuder** [comment]
Hi bot, please keep this PR open. //cc @ldaley

**Comment by KirovClassBattlecruiser** [comment]
Why hasn't this been merged? It's just a slight improvement sure but there is no reason to not merge this

**Comment by ldaley** [comment]
These changes in and of themselves don't provide enough value to justify the change. They have effectively been superseded by intentions for more fundamental changes that should land over the coming months.

**Pull request closed without merging**